[MM-19740] Return error for non-system-admins/team admins/channel admins from the GET /api/v4/groups API endpoint response

hahmadia opened the pull request:
#### Summary
**UPDATE November 5, 2019**

Following up discussions with @ali-farooq0 and @mkraft, the decision was made that you must always pass a `not_associated_to_channel` or `not_associated_to_team` param to `GET /groups` and it is checked to see if you have permissions to manage that particular channel (_PERMISSION_MANAGE_PRIVATE_CHANNEL_MEMBERS_ or _PERMISSION_MANAGE_PUBLIC_CHANNEL_MEMBERS_) or view that team (_PERMISSION_MANAGE_TEAM_). If the above is not satisfied, an error is returned otherwise the groups are returned with all their fields.

It's been confirmed that in the Matermost-Redux repo, when we do a call to `GET /groups` that we are always passing a `not_associated_to_team` or `not_associated_to_channel` parameter. 

#### Ticket
https://mattermost.atlassian.net/browse/MM-19740

ali-farooq0 commented on model/group.go:
```diff
@@ -43,6 +43,14 @@ type Group struct {
 	MemberCount  *int        `db:"-" json:"member_count,omitempty"`
 }
 
+type GroupNonAdmin struct {
```
not sure why we need to create another struct for this, can we not just sanitize the values when we get groups back? i.e. sanitize them for non-admins. We can include the `omitempty` tag for fields we want to remove. Thoughts?

ali-farooq0 replied:
Spoke to @hahmadia, regarding why we can't use `omitempty` as for the regular case we want to include fields with default values.

Suggesting to use [anonymous struct](https://www.geeksforgeeks.org/anonymous-structure-and-field-in-golang/) for marshalling purposes since it'll _only_ be used for marshalling, better t o not create a special struct that's a first class entity as opposed to a anonymous struct only used for marshalling.

hahmadia replied:
@ali-farooq0 That's a great point and it totally makes sense why it would be better to use anonymous struct rather than creating a new struct in our model file. I have made the changes to reflect your suggestion. Thanks! 😄

hahmadia commented on api4/group.go:
```diff
@@ -616,7 +616,25 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
+		b, marshalErr = json.Marshal(groupsNonAdmin)
+	} else {
+		b, marshalErr = json.Marshal(groups)
+
```
Fixed!

hahmadia commented on api4/group.go:
```diff
@@ -616,7 +616,25 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 
-	b, marshalErr := json.Marshal(groups)
+	var b []byte
```
Fixed!

hahmadia commented:
@mkraft I have added the test case as per your suggestion. I just omitted checking the `hasSyncables` field since it doesn't change  (It will always be false as it doesn't get set by calling `GET /groups`).

ali-farooq0 commented on api4/group_test.go:
```diff
@@ -8,12 +8,67 @@ import (
+	}
+
+	a, _ := th.Client.GetGroups(opts)
+	var nonAdminGroup, adminGroup model.Group
```
@hahmadia 
Ahh I totally forgot that we had Client4 API to get groups. In this case don't we have the exact same issue where we're returning _all_ the fields and the client has to figure out if a particular value is a real value or the default value which they're getting because they're non-admins. Hmmm... Does non-admins need to ever use this API? If not I would suggest to throw an unauthorized error, and return no data.

ali-farooq0 replied:
apologies for missing this part in the first pass 🤦‍♂

hahmadia commented:
Following up discussions, the decision was made that you must always pass a `not_associated_to_channel` or `not_associated_to_team` param to `GET /groups` and it is checked to see if you have permissions to manage that channel or team. If you don't, then an error is returned, otherwise all fields are returned to you.

It's been confirmed that in the Matermost-Redux repo, when we do a call to `GET /groups` that we are always passing a `not_associated_to_team` or `not_associated_to_channel` parameter.

amyblais commented:
@thefactremains @hahmadia Why is this assigned to v5.17?

hahmadia commented:
> @thefactremains @hahmadia ~Why is this assigned to v5.17?~ Was added by mistake

Sorry about that. A small error on our part. Has been updated in Jira

ali-farooq0 commented on api4/group.go:
```diff
@@ -610,6 +609,11 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
 		opts.NotAssociatedToChannel = channelID
 	}
 
+	if getTeamErr != nil && getChannelErr != nil {
```
this should be moved up as suggested above.

hahmadia replied:
Thanks! Done.

ali-farooq0 commented on api4/group.go:
```diff
@@ -575,28 +575,27 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
+	teamID := c.Params.NotAssociatedToTeam
+	channelID := c.Params.NotAssociatedToChannel
```
This method requires either `teamID` or `channelID` right? or both? could just do something like if either is required
```
	if len(teamID) != 26 && len(channelID) != 26 {
		c.Err = model.NewAppError("Api4.getGroups", "api.ldap_groups. invalid_or_missing_channel_or_team_id", nil, "", http.StatusBadStatus)
		return
	}
```

hahmadia replied:
Thanks! Done.

ali-farooq0 commented on api4/group.go:
```diff
@@ -575,28 +575,27 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 
-	teamID := c.Params.NotAssociatedToTeam
-	if len(teamID) == 26 {
```
and this can just be `len(teamID) > 0`

hahmadia replied:
Thanks! Done.

hahmadia commented on api4/group.go:
```diff
@@ -575,28 +575,27 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
-	if len(teamID) == 26 {
+	// Check to see if valid TeamID has been provided, then check to see if the user requesting this has
+	// permission to the team in order to add the teamID to the opts
+	_, getTeamErr := c.App.GetTeam(teamID)
```
Thanks! Done.

hahmadia commented on api4/group.go:
```diff
@@ -575,28 +575,27 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
-	teamID := c.Params.NotAssociatedToTeam
-	if len(teamID) == 26 {
+	// Check to see if valid TeamID has been provided, then check to see if the user requesting this has
+	// permission to the team in order to add the teamID to the opts
```
Thanks! Done.

hahmadia commented on api4/group.go:
```diff
@@ -575,28 +575,27 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
-			c.Err = err
-			return
-		}
+	channel, getChannelErr := c.App.GetChannel(channelID)
```
Thanks! Done.

mkraft commented on api4/group.go:
```diff
@@ -575,23 +575,33 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
 		c.Err = model.NewAppError("Api4.getGroups", "api.ldap_groups.license_error", nil, "", http.StatusNotImplemented)
 		return
 	}
+	teamID := c.Params.NotAssociatedToTeam
```
```suggestion
	var teamID string
	if id := c.Params.NotAssociatedToTeam; model.IsValidId(id) {
	    teamID = id
	}
```

mkraft replied:
Just a suggestion so that you know if these variables are non-default values that they're valid Mattermost IDs formats. 

The way it is currently written you can get down and inside the branches to make the `GetTeam` or `GetChannel` method calls with any non-zero-length string.

mkraft commented on api4/group.go:
```diff
@@ -575,23 +575,33 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
+	teamID := c.Params.NotAssociatedToTeam
+	channelID := c.Params.NotAssociatedToChannel
```
```suggestion
	var channelID string
	if id := c.Params.NotAssociatedToChannel; model.IsValidId(id) {
	    channelID = id
	}
```

mkraft commented on api4/group.go:
```diff
@@ -575,23 +575,33 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
+	teamID := c.Params.NotAssociatedToTeam
+	channelID := c.Params.NotAssociatedToChannel
+
+	if len(teamID) <= 0 && len(channelID) <= 0 {
```
```suggestion
	if teamID == "" && channelID == "" {
```

mkraft commented on api4/group.go:
```diff
@@ -575,23 +575,33 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
 
-	teamID := c.Params.NotAssociatedToTeam
-	if len(teamID) == 26 {
+	if len(teamID) > 0 {
```
```suggestion
	if teamID != "" {
```

mkraft commented on api4/group.go:
```diff
@@ -575,23 +575,33 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
 
-	channelID := c.Params.NotAssociatedToChannel
-	if len(channelID) == 26 {
+	if len(channelID) > 0 {
```
```suggestion
	if channelID != "" {
```

mkraft commented on api4/group.go:
```diff
@@ -575,23 +575,33 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
+			c.Err = err
+			return
+		}
 		if !c.App.SessionHasPermissionToTeam(c.App.Session, teamID, model.PERMISSION_VIEW_TEAM) {
```
I thought the plan was to change this permission? Or am I wrong? Obviously this would need to be tested/verified. The intent is to restrict it to just team-admins.
```suggestion
		if !c.App.SessionHasPermissionToTeam(c.App.Session, teamID, model.PERMISSION_MANAGE_TEAM) {
```

ali-farooq0 commented on api4/group.go:
```diff
@@ -575,23 +575,33 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
+	teamID := c.Params.NotAssociatedToTeam
+	channelID := c.Params.NotAssociatedToChannel
+
+	if len(teamID) <= 0 && len(channelID) <= 0 {
```
hmm can we just instead do the following?

```
teamID := c.Params.NotAssociatedToTeam
channelID := c.Params.NotAssociatedToChannel

if !model.IsValidId(teamID) && !model.IsValidId(channelID) {
c.Err = model.NewAppError("Api4.getGroups", "api.getGroups.invalid_or_missing_channel_or_team_id", nil, "", http.StatusBadRequest)
		return
	}

```

This would simplify the validation logic above.

mkraft replied:
That only tells you if both parameters are invalid, but you’d still have to invoke `IsValidId` again in each of the branching statements below to know whether to invoke `GetChannel` and `GetTeam`. With my proposal you can just test if they’re empty strings or not.

ali-farooq0 replied:
Ahh fair enough. Makes sense. 👍 to your suggestions.

hahmadia commented:
@mkraft @ali-farooq0 
Thank you for the review. I have addressed the stylistic and permission issues brought up. I also fixed the test cases. Ready for re-review 😺

ali-farooq0 commented on i18n/en.json:
```diff
@@ -1340,6 +1340,10 @@
   },
+  {
+    "id": "api.getGroups.invalid_or_missing_channel_or_team_id",
+    "translation": "Invalid/Missing channel ID or Team ID"
```
non blocking nit: missing `.` in the end.

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-server-pr-12961.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server destroyed

hahmadia commented:
@srkgupta 
Upon reviewing the problem, the cause of the problem is most likely due to your permissions scheme allowing all users to manage channel members. If you disable this, then you should get a permissions error.
